### PR TITLE
Readonly properties and index signatures

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR adds support for readonly properties and index signatures as suggested in #12. Changes include:

* A new `readonly` modifier that can applied to property and index signature declarations.
* A property or index signature declared with the `readonly` modifier is considered read-only.
* Read-only properties may have initializers and may be assigned to in constructors within the same class declaration, but otherwise assignments to read-only properties are disallowed.
* Assignments to read-only index signatures are always disallowed.

Entities are *implicitly* read-only in several situations:

* A property declared with a `get` accessor and no `set` accessor is considered read-only.
* In the type of an enum object, enum members are considered read-only properties.
* In the type of a module object, exported `const` variables are considered read-only properties.
* An entity declared in an `import` statement is considered read-only.
* An entity accessed through an ES2015 namespace import is considered read-only (e.g. `foo.x` is read-only when `foo` is declared as `import * as foo from "foo"`).

Some examples:

```typescript
interface Point {
    readonly x: number;
    readonly y: number;
}

var p1: Point = { x: 10, y: 20 };
p1.x = 5;  // Error, p1.x is read-only

var p2 = { x: 1, y: 1 };
var p3: Point = p2;  // Ok, read-only alias for p2
p3.x = 5;  // Error, p3.x is read-only
p2.x = 5;  // Ok, but also changes p3.x because of aliasing
```
```typescript
class Foo {
    readonly a = 1;
    readonly b: string;
    constructor() {
        this.b = "hello";  // Assignment permitted in constructor
    }
}
```
```typescript
// Read-only view of Array<T> with all mutating methods removed
interface ReadonlyArray<T> {
    readonly [x: number]: T;
    readonly length: number;
    toString(): string;
    toLocaleString(): string;
    concat<U extends ReadonlyArray<T>>(...items: U[]): T[];
    concat(...items: T[]): T[];
    join(separator?: string): string;
    slice(start?: number, end?: number): T[];
    indexOf(searchElement: T, fromIndex?: number): number;
    lastIndexOf(searchElement: T, fromIndex?: number): number;
    every(callbackfn: (value: T, index: number) => boolean, thisArg?: any): boolean;
    some(callbackfn: (value: T, index: number) => boolean, thisArg?: any): boolean;
    forEach(callbackfn: (value: T, index: number) => void, thisArg?: any): void;
    map<U>(callbackfn: (value: T, index: number) => U, thisArg?: any): U[];
    filter(callbackfn: (value: T, index: number) => boolean, thisArg?: any): T[];
    reduce(callbackfn: (previousValue: T, currentValue: T, currentIndex: number) => T, initialValue?: T): T;
    reduce<U>(callbackfn: (previousValue: U, currentValue: T, currentIndex: number) => U, initialValue: U): U;
    reduceRight(callbackfn: (previousValue: T, currentValue: T, currentIndex: number) => T, initialValue?: T): T;
    reduceRight<U>(callbackfn: (previousValue: U, currentValue: T, currentIndex: number) => U, initialValue: U): U;
}

let a: Array<number> = [0, 1, 2, 3, 4];
let b: ReadonlyArray<number> = a;
b[5] = 5;      // Error, elements are read-only
b.push(5);     // Error, no push method (because it mutates array)
b.length = 3;  // Error, length is read-only
a = b;         // Error, mutating methods are missing
```

Note that the `readonly` modifier does not affect subtype and assignability relationships of the containing type for the [reasons described here](https://github.com/Microsoft/TypeScript/pull/6532#issuecomment-174356151). Also note that this PR specifically doesn't attempt to formalize immutable objects. However, with the `readonly` modifier it becomes a lot more convenient to implement such objects.

The PR still needs tests, but I thought I'd get it up there for folks to take a look at.

### CODE REVIEW

**Comment by JsonFreeman** [comment]
Yay!

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -134,6 +134,8 @@ namespace ts {
         const anySignature = createSignature(undefined, undefined, emptyArray, anyType, 0, /*hasRestParameter*/ false, /*hasStringLiterals*/ false);
         const unknownSignature = createSignature(undefined, undefined, emptyArray, unknownType, 0, /*hasRestParameter*/ false, /*hasStringLiterals*/ false);
 
+        const enumNumberIndexInfo = createIndexInfo(stringType, /*isReadonly*/ false);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: What is this?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -3620,29 +3612,34 @@ namespace ts {
...
+            const numberIndexInfo = getUnionIndexInfo(type.types, IndexKind.Number);
+            setObjectTypeMembers(type, emptySymbols, callSignatures, constructSignatures, stringIndexInfo, numberIndexInfo);
         }
 
         function intersectTypes(type1: Type, type2: Type): Type {
             return !type1 ? type2 : !type2 ? type1 : getIntersectionType([type1, type2]);
         }
 
+        function intersectIndexInfos(info1: IndexInfo, info2: IndexInfo): IndexInfo {
+            return !info1 ? info2 : !info2 ? info1 : createIndexInfo(
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: `info1 || info2 || createIndexInfo(...`?

File: src/compiler/checker.ts (hunk 1 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Why is this indexer not readonly?

File: src/compiler/diagnosticMessages.json [hunk 3]
```diff
@@ -987,6 +995,10 @@
         "category": "Error",
         "code": 2342
     },
+    "Property '{0}' is read-only in type '{1}' but read-write in type '{2}'.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: "but writable in type '{2}'."

File: src/compiler/checker.ts (hunk 1 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: I believe because we don't want to break this code:
```ts
enum SyntaxKind { Node, OtherNode }
function dbgPrint(value: number, yourEnum: { [n: number]: string }) {
	console.log(yourEnum[value]);
}
// Would-be error, dbgPrint wants mutable index signature but enum has readonly signature
dbgPrint(SyntaxKind.OtherNode, SyntaxKind);
```

Comment (JsonFreeman) [inline src/compiler/checker.ts]: Ah right, because of the assignability rule. Makes sense.

Comment (ahejlsberg) [inline src/compiler/checker.ts]: That's right. It's bit of a puzzle ensure backwards compatibility, so there are actually two forms of read-only-ness in the new implementation:

* `readonly` properties, `readonly` index signatures, and `get` only accessors are *always* considered read-only, in both assignments and type compatibility checks.
* `const` variables, functions, classes, namespaces, enums, and enum members are considered read-only in assignments, *but not* in type compatibility checks. This ensures backwards compatibility. For example, it is reasonably common to implement an interface contract using a module or namespace, and strict read-only-ness would break that pattern.

File: tests/baselines/reference/decrementOperatorWithEnumType.errors.txt [hunk 4]
```diff
@@ -1,21 +1,27 @@
...
 
-==== tests/cases/conformance/expressions/unaryOperators/decrementOperator/decrementOperatorWithEnumType.ts (3 errors) ====
+==== tests/cases/conformance/expressions/unaryOperators/decrementOperator/decrementOperatorWithEnumType.ts (5 errors) ====
     // -- operator on enum type
     
     enum ENUM1 { A, B, "" };
     
     // expression
     var ResultIsNumber1 = --ENUM1["A"];
+                            ~~~~~~~~~~
```
Comment (yuit) [inline tests/baselines/reference/decrementOperatorWithEnumType.errors.txt]: As clarification, increment/decrement operator and any assignment to enum isn't allowed as enum type is read-only upon assignment? if that is the case, I think the error message should be updated.

File: src/compiler/checker.ts (hunk 1 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I see. That sounds like a good distinction.

**Comment by joelday** [comment]
Just stumbled upon this. Don't think I could be any happier than I am right now.

**Comment by ahejlsberg** [comment]
Latest commits reflect our design meeting discussions. Most significant change is that, in order to ensure backwards compatibility, the `readonly` modifier no longer affects subtype and assignability type relationships of the containing type (but of course it still affects assignment operations).

Consider the following code:

```typescript
interface ArrayLike<T> {
    length: number;
    [index: number]: T;
}

function foo(array: ArrayLike<string>) {
    // Doesn't mutate array
}

var s = "hello";
var a = ["one", "two", "three"];
foo(s);  // s has readonly length and index signature
foo(a);
```

In existing TypeScript code there is no way to indicate if a particular property is intended to be read-only or mutable. In the code above, `foo` doesn't mutate the array it is passed, but there is nothing in the code that says it can't. However, now that we've added a `readonly` modifier to the `length` property and the index signature in the `String` interface (because they really are read-only), the `foo(s)` call above would be an error if we said that a `readonly` property is incompatible with a property without `readonly`. Specifically, we can't interpret the *absence* of a `readonly` modifier to mean read-write, we can only say that we don't know. So, if an interface differs from another interface only in the `readonly` modifiers on its properties, we have to say that the two interfaces are compatible. Anything else would be a massive breaking change.

This change isn't as bad as it might sound because we'll always error on direct assignments to `readonly` properties and that's where the majority of the errors occur. We're still thinking about possible ways to further describe and check the mutation semantics of types, but that is beyond this particular PR.

**Comment by ahejlsberg** [comment]
@mhegazy @RyanCavanaugh I think this one is ready to merge if you want to take a look.

File: src/compiler/types.ts [hunk 5]
```diff
@@ -369,32 +370,33 @@ namespace ts {
...
-        HasParamDecorators =    1 << 24,  // If the file has parameter decorators (initialized by binding)
-        HasAsyncFunctions =     1 << 25,  // If the file has async functions (initialized by binding)
+        None =               0,
+        Export =             1 << 1,  // Declarations
+        Ambient =            1 << 2,  // Declarations
+        Public =             1 << 3,  // Property/Method
+        Private =            1 << 4,  // Property/Method
+        Protected =          1 << 5,  // Property/Method
+        Static =             1 << 6,  // Property/Method
+        Readonly =           1 << 7,  // Property/Method
```
Comment (RyanCavanaugh) [inline src/compiler/types.ts]: It looks like `1 << 0` is available (?) if you want to avoid renumbering the whole enum.

**Comment by RyanCavanaugh** [comment]
:+1:

Any thoughts on bumping this into 1.8 beta?

**Comment by simonbuchan** [comment]
Any chance for `readonly interface Foo {}` / `readonly class Foo {}` implying `readonly` on all fields, given the obvious use of this for immutable types?
Not sure what that should do with setter members or non-readonly members of extended / implemented types, though.

**Comment by ahejlsberg** [comment]
@simonbuchan We're not going to support `readonly` on interface and class declarations. However, we're thinking about a `readonly` type modifier that could be applied to *any* type, as in:

```typescript
let x: readonly string[];
let y: readonly { names: string[], values: number[] };
```

For a given type `T` the effect of `readonly T` would be:

* Change each property `p: X` to `readonly p: readonly X` (i.e. make it deeply readonly).
* Make it an error to invoke mutating methods of `T`.

The default assumption would be that all methods and set accessors are mutating and all get accessors are non-mutating, but it would be possible to add a modifier (don't have a good name for it yet) to a method have it be considered non-mutating (it's not clear that we can do this analysis automatically, and even if we could, it is not uncommon to have methods that mutate internal caches but want to be considered non-mutating from the outside).

Anyway, this is way beyond the scope of this PR, but we're going to hold off on further uses of the `readonly` modifier until we have it all worked about.

**Comment by simonbuchan** [comment]
> add a modifier (don't have a good name for it yet) to a method have it be considered non-mutating

Hey, you got your C++ in my JavaScript! But I like this plan :+1:

Would the idea be that `ReadOnlySomeCollection<T>` etc... would be obsolete?

**Comment by ahejlsberg** [comment]
@simonbuchan Yes, idea is to automatically synthesize read-only views of arbitrary types.

**Comment by basarat** [comment]
> However, we're thinking about a readonly type modifier that could be applied to any type

I would imagine something that allows the following transparency : 

```ts
/** Tell the consumer that we will not modify config in any way */
function foo(config: { someData: any }) {
    // prevent the following:
    config.someData = config.someData || '123';
}
```

**Comment by myitcv** [comment]
@ahejlsberg thanks very much for this change. Working really well for us in the latest nightly build!

**Comment by myitcv** [comment]
@ahejlsberg - just as a quick follow-up, it looks like formatting of interface defintions that contain the `readonly` modifier are broken. The following:

```typescript
interface ExampleInt {
	readonly FirstVal: number;
	readonly SecondVal: number;
	readonly ThirdVal: number;
	Resolve(n: number): number;
}
```

gets reformatted (via `tsfmt`) as:

```typescript
interface ExampleInt {
	readonly FirstVal: number;
readonly SecondVal: number;
readonly ThirdVal: number;
Resolve(n: number): number;
}
```

**Comment by RyanCavanaugh** [comment]
@myitcv can you log a separate issue please?

**Comment by myitcv** [comment]
@RyanCavanaugh - [done](https://github.com/Microsoft/TypeScript/issues/6856)

**Comment by jpablo** [comment]
I was wondering why not use the keyword `const` instead of adding a new keyword.

**Comment by danquirk** [comment]
Re-using keywords for similar but different purposes can be dangerous (whether that keyword carries baggage from this language or some other languages). For example there has been a fair amount of confusion around ```typeof``` with where it does its JavaScript thing and where it does its TypeScript thing and what the differences are.

**Comment by jpablo** [comment]
I guess I misunderstood the purpose of the keyword. I assumed it was to prevent re-assignment, in which case a single keyword seems enough (as exemplified by the following Scala code):
```
  trait A { val x: Int }
  val a = new A { val x: Int = 1 }
  // a = ... // Illegal
  // a.x = ... // Illegal too
```

**Comment by basarat** [comment]
I love this feature :heart:  (and wrote about it too https://basarat.gitbooks.io/typescript/content/docs/types/readonly.html) 

But I do feel that `const` would have been better. Here is how a const behaves: 

```ts

namespace MustInit {
    const foo: { bar: number }; // Error: const declaration must be initialized (1)
}

namespace ImmutableAtRoot {
    const foo = {bar : 123};
    foo = {bar:456}; // Error: left hand side cannot be `const` or `readonly`
}

namespace NotDeeplyImmutable {
    const foo = {bar : 123};
    foo.bar = 456; // Okay
}
```
Very similar to `readonly`:
```ts
namespace MustInit {
    class Foo {
        readonly foo: {bar: number};
    }
    // Foo.foo is farily useless uninitialized :-/
}

namespace ImmutableAtRoot {
    class Foo {
        readonly foo = { bar: 123};
    }
    new Foo().foo = {bar: 456}; // Error: left hand side cannot be `const` or `readonly`
}

namespace NotDeeplyImmutable {
    class Foo {
        readonly foo: {bar: number};
    }
    new Foo().foo.bar = 456; // Okay
}
```




I would imagine using `const` might help with future ES providing this as an option on initialized properties. Also not using `readonly` might help preserve the keyword for deep immutability (https://github.com/Microsoft/TypeScript/pull/6532#issuecomment-175273557) instead of shallow immutability like `const`.

Of course your constraints might be ESNext alignment or something else that I'm missing. I still love the attention you guys are paying to deterministic / maintainable code bases :rose:

**Comment by RyanCavanaugh** [comment]
`const` is not the correct analogue because a `const` binding cannot be re-assigned; this invariant is *not* true of `readonly` properties due to the possibility of aliasing.

Consider:
```ts
const a = /* any expression */;
const q = a;
const p = q;

/* run *any code imaginable* here */

// Always prints 'true'
console.log(p === q);
```
vs this:
```ts
interface ImmutableHolder {
    readonly x: any;
}

const a = { x: 42 };
const b: ImmutableHolder = a;
const c = b.x;

/* easily imagined code follows */
a.x = 100;

// Prints 'false'
console.log(a.x === c);
```

**Comment by aluanhaddad** [comment]
This is beautiful. I can't wait for TypeScript 2.0. Great work!

**Comment by basarat** [comment]
Ahh  readonly member is actually more mutable than const. Thanks for the clarification 🌹😌

**Comment by ahejlsberg** [comment]
A good way to think of it is:
* `readonly` = cannot be modified by me
* `const` = cannot be modified, period

**Comment by Robert923** [comment]
Will "readonly" work as a constructor parameter property?

`constructor( readonly x : number ) {
...
}`

**Comment by RyanCavanaugh** [comment]
@Robert923 not currently. You can log a new suggestion if you'd like.

General policy so far has been that we don't like to allow a new modifier *everywhere* and instead let it in only where clearly needed (to avoid having "modifier soup" constructs like `export readonly public default const internal enum Foo { ... }`)

**Comment by electricessence** [comment]
Oh man.  Is there something like Object.freeze for a single property?  That would be a nice add.

**Comment by JsonFreeman** [comment]
@electricessence You use a property descriptor for your property, you can set the writable to false.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty

**Comment by malibuzios** [comment]
@ahejlsberg [These questions](https://github.com/Microsoft/TypeScript/issues/8496#issuecomment-217618696) in #8496 (and the issue in general) are highly relevant to the topic here. Consider providing your comments. I have only closed that issue because I felt that your (unbelievably stubborn and righteous) team members would somehow find it more 'respectful' that I show my complete pessimism at having any influence here than trying to 'confront' them in any way. **You** have already won the 'battle', congratulations, now feel free to join the discussion.

**Comment by joelday** [comment]
Unbelievably stubborn and righteous, eh?

**Comment by malibuzios** [comment]
[I apologize for this being very off-topic but since I was asked I felt I had no choice but answering here, the person who asked doesn't have a public email so there is no way for me to contact them personally]

@joelday 

Yes, my personal impression is that they are very 'difficult' people to deal with, in the sense that they are putting as much energy into 'justifying' (sometimes dishonestly or irrationally) their actions and decisions to actually considering things openly and objectively when presented to them. They even do this on some very trivial subjects (e.g. you tell them they don't conform to the ES6 spec by [practically disallowing inheritance of static accessors for ES5 targets](https://github.com/Microsoft/TypeScript/issues/1520) and instead of listening and acknowledging the problem like a normal person would do, a group of team members just 'jumps' on you and starts to throw some strange and arbitrary ('canonical') justifications about performance consideration and dubious statistics there is '0%' anyone would ever want to do this.. well I did! and so did others who reported this problem, and probably many who didn't.. :( ).

They put you in a situation where you have to be so eloquent and rational, to the point you have to literally spend dozens of hours into very carefully and rationally crafting a 'thesis' for something otherwise relatively trivial, just for them. And they on the other side are allowed to simply 'push' their opinions and twisted patterns of 'justification' arbitrarily without anyone saying anything. When you start challenging those patterns they would either ignore you or tell you you are intentionally 'wasting their time' with things that aren't important. They would also use their 'internal' understanding of the language and compiler by mentioning intricate things or concepts you don't know or completely understand and dishonestly use that against you.

They take submitted ideas they like but they don't actually partake in a real two-way conversations about it. They don't really involve the submitter of their personal ideas and insights, at least _before_ they decide on their actions. And when/if they do inform back, they would condense it to a set of a few sentences that are not very useful or enlightening by themselves (or direct you to a set of design 'notes' that are very hard to extract anything useful from). You can't ask a (non-trivial) question because that's just 'wasting their time' (they don't do 'one on one').

If they don't like an idea, they would usually 'tank' and attack it mercilessly to the point of almost (or perhaps actually) insulting the person who submitted it. They almost never 'improve' on things suggested to them, at least not publicly. They would rather spend their time and 'smarts' in crafting careful arguments for justifying why they think someone are 'wrong' and they must be 'right' than taking a step back to seriously and objectively reconsider things, and eventually thank the participant for contributing something useful to the discussion.

This has been my experience so far, anyway. I'm mostly referring here to the 2-3 members who reply to GitHub issues.

**Comment by RReverser** [comment]
Is there any reason / chance to change that the generated code contains regular property assignment and not `defineProperty` with `writable: false`? That would help a lot to protect against runtime assignment to properties designed as `readonly`.

**Comment by Robert923** [comment]
I agree with @RReverser - on my team we frequently use TypeScript compiled code from (legacy)JavaScript applications - so keeping as many of those protections as possible would be very helpful.

**Comment by Aides359** [comment]
Is this already available or only for TypeScript 2.0 (neither VS nor TypeScript Playground seem to understand it)

**Comment by RReverser** [comment]
@Aides359 As usual with all the fresh stuff, you need to use `npm i typescript@next` to get the latest compiler with experimental features included.

**Comment by aluanhaddad** [comment]
@Aides359 I think the playground hosts the latest stable release. I recommend doing what @RReverser suggests and additionally installing vscode and setting then `"typescript.tsdk"` property of your user/workspace settings to the install location's lib directory.

**Comment by paleo229** [comment]
> Read-only properties may have initializers and may be assigned to in constructors within the same class declaration, but otherwise assignments to read-only properties are disallowed.

Why disallow assignments?

A `readonly` property would be useful to implement getter with a true property (and better performance):

    class Foo {
        readonly a = 1;
        inc() {
            ++this.a;  // Why not allow to do this?
        }
    }

Does the limitation exist for a technical reason?

**Comment by RyanCavanaugh** [comment]
If that were allowed we'd have daily reports about "Bug: Why does TS let me mutate `readonly` fields?" :wink:

We discussed the notion of "public readonly, protected mutable" but opted not to simply due to complexity. You can already have a getter-only wrapper around a private variable, which achieves the same semantics at the cost of slightly increased class size.

**Comment by paleo229** [comment]
OK. You use the term `readonly` as a synonym for `const`. I hoped an _accessor_ `readonly`. Too bad for me. :(

**Comment by aluanhaddad** [comment]
> OK. You use the term readonly as a synonym for const. I hoped an accessor readonly. :(

@paleo229
A property with no `set` attribute is `readonly` by implication you so don't need a modifier at all.
```typescript
class Person {
    get name(): string {
        return 'Smith';
    }
}

let smith = new Person();
smith.name = 'Boyde'; // error
```

**Comment by Attrash-Islam** [comment]
I use this great feature for mutable preventing by the syntax highlighting. Thanks a lot for this! 

+++

**Comment by aleksey-bykov** [comment]
> This change isn't as bad as it might sound because we'll always error on direct assignments to readonly properties and that's where the majority of the errors occur. 

this is a horrible change, rendering this feature usless

**Comment by aleksey-bykov** [comment]
simple use case that fails, I have my app about 1500 TS files thick, in this app I have an interface named Data that was intended to be immutable historically could not be forced being so

With the readonly feature announced it was immediately clear that such enforcment can be imposed.

However it turned out that there were a few places where Data was meant to be mutable.

To address it the original Data interface was split into ReadOnlyData and WitableData, however it didn't help much due to the said decision. Details can bee seen in #13000.

Bottom line is we are still not sure if Data is effectively mutable or not, despite applibg `readonly` to all its properties.

**Comment by masaeedu** [comment]
@ahejlsberg Correct me if I'm wrong, but the only reason interface compatibility predicated on readonly modifiers is a breaking change is because of readonly being implicitly applied to everything. The tradeoff of worse soundness in return for saving the user some typing, especially when you now have mapped types, is not a good one IMO.
